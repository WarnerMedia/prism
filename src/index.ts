// Copyright (c) Warner Media, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
import { Psm, PsmConfig } from "./Psm";

const psm = new Psm();

export * from "./Psm";
declare global {
  interface Window {
    psmMgrConfig: PsmConfig;
    psmMgr: {
      psm_test_ts: boolean;
      psmEnvironment: string;
      psmSendGeoInit: boolean;
      psmHost: string;
      psmGeoActualCookieValue: string;
    };
    psmCore?: { psm?: { browser?: string } };
  }
  interface Crypto {
    randomUUID: () => string;
  }
}

window.psmMgr = {
  ...window.psmMgr,
  psm_test_ts: true, // FIXME: for dev purposes, allows our logger function to log to the console
};
