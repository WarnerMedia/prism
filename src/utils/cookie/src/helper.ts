import { CookieOptions } from ".";

export const debug = (...args: (string | object | number | boolean)[]) => {
  if (window.location.search.search(/[?&]psm_debug=[1t]/) !== -1) {
    console.info(...args);
  }
};

export function cookieHelper(
  name?: string,
  value?: string,
  options?: Partial<CookieOptions>
): string | object | void {
  switch (arguments.length) {
    case 3:
    case 2:
      return set(name, value, options);
    case 1:
      return get(name);
    default:
      return all();
  }
}

/**
 * Set cookie `name` to `value`.
 */
function set(name: string, value: string, options: Partial<CookieOptions>) {
  options = options || {};
  let str = encode(name) + "=" + encode(value);

  if (options.encode === false) {
    str = name + "=" + value;
  }

  if (null == value) options.maxage = -1;

  if (options.maxage && !options.expires) {
    options.expires = new Date(+new Date() + options.maxage);
  }

  if (options.path) str += "; Path=" + options.path;
  if (options.domain) str += "; Domain=" + options.domain;
  if (options.expires) str += "; Expires=" + options.expires.toUTCString();
  if (options.samesite) str += "; SameSite=" + options.samesite;
  if (options.secure) str += "; Secure";

  debug("[COOKIE]: cookie.helper.set()", str, options);

  document.cookie = str;
}

/**
 * Return all cookies.
 */
function all(): object {
  let str: string;
  try {
    str = document.cookie;
  } catch (err) {
    if (typeof console !== "undefined" && typeof console.error === "function") {
      console.error(err.stack || err);
    }
    return {};
  }
  return parse(str);
}

/**
 * Get cookie `name`.
 */
function get(name: string): string {
  const val = all()[name];

  debug(`[COOKIE]: helper.get() - getting ${name} cookie value ${val}`);

  return val;
}

/**
 * Parse cookie `str`.
 */
function parse(str: string): { [key: string]: string } {
  const obj = {};
  const pairs = str.split(/ *; */);
  let pair: string[];

  if ("" == pairs[0]) return obj;

  for (let i = 0; i < pairs.length; ++i) {
    const cookieName = pairs[i].substr(0, pairs[i].indexOf("="));
    const cookieValue = pairs[i].substr(pairs[i].indexOf("=") + 1);
    obj[decode(cookieName)] = decode(cookieValue);
  }

  return obj;
}

function encode(value: string | number | boolean): string {
  try {
    return encodeURIComponent(value);
  } catch (err) {
    // ignore
  }
}

function decode(value: string): string {
  try {
    return decodeURIComponent(value);
  } catch (err) {
    // ignore
  }
}
