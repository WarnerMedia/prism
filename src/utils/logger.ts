/* eslint-disable @typescript-eslint/no-explicit-any */
import { IQueryFeatureResult } from "../Psm";
import Queue from "@segment/localstorage-retry";
import { URLS, queueOptions } from "./constants";
import { hasLocalStorage } from "./browser";
import { sendRequest, SendRequestError } from "./sendRequest";
import { Psm } from "../Psm";
import pkg from "../../package.json";

const win = window;
let logger: Queue;
let environment: string;
let values: Psm;

export interface LogOptions {
  flgs?: IQueryFeatureResult[];
  err?: Error | SendRequestError;
  methodName: string;
  eventType: string;
  ukid?: string;
  context?: object;
  message?: string;
}

export function createLogger(psm: Psm) {
  environment = psm.config.psmEnvironment
    ? psm.config.psmEnvironment.toUpperCase()
    : "PROD";
  values = psm;

  logger = new Queue("logger", queueOptions, (item: any, done: any) => {
    if (/bot|crawl|spider/i.test(win.navigator.userAgent)) {
      return done(null, {});
    }
    item.ts = new Date().toISOString();
    sendRequest(URLS.logs[environment], {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(item),
    })
      .then((res) => done(null, res))
      .catch((err) => done(err));
  });

  logger.start();
}

/**
 * Not intended to be called directly. Only to be invoked by debug(), warn(), and error()
 * and only exported for testing purposes.
 */
export function log(options: LogOptions, level: string) {
  let message: string;
  let stack: string;
  let additionalContext: object;

  if (options.err) {
    message = options.err.toString();
    stack = options.err.stack;
    additionalContext = options.err["context"] || {};
  } else {
    message = options.message;
    additionalContext = options.context;
  }

  const item = {
    brand: values.config.brand,
    device: {
      type: win.navigator.platform,
      userAgent: win.navigator.userAgent,
    },
    eventName: level,
    eventProperties: {
      logMessage: message,
      method: options.methodName,
      featureFlagValues: values.getFlags(),
      stackTrace: stack,
    },
    eventType: "log",
    eventTimestamp: new Date().toISOString(),
    library: {
      name: "PrismJS",
      version: pkg.version,
      initConfig: values.config,
    },
    platform: "web",
    session: values.session, // cannot update session values here, as it creates an infinite loop!
    ukid: values.ukid,
  };

  const localStorageDebugFlag =
    hasLocalStorage() && win.localStorage.getItem("psm_debug") == "true";
  const queryStringDebugFlag = win.location.search.includes("psm_debug");

  if (level !== "debug") {
    if (
      localStorageDebugFlag ||
      queryStringDebugFlag ||
      environment === "AUTOMATED_TEST"
    ) {
      if (environment === "AUTOMATED_TEST") {
        console[level](
          `[PSM]: ${level}:`,
          JSON.stringify(item.eventProperties)
        );
      } else {
        console[level](`[PSM]: ${level}:`, item);
      }
    } else {
      if (item.eventProperties.featureFlagValues["doppler-send-logs"]) {
        logger.addItem(item);
      }
    }
  } else {
    if (
      localStorageDebugFlag ||
      queryStringDebugFlag ||
      environment === "AUTOMATED_TEST"
    ) {
      if (environment === "AUTOMATED_TEST") {
        // console.log(`[PSM]: DEBUG`, JSON.stringify(item.eventProperties));
      } else {
        // console.log(`[PSM]: DEBUG`, item);
      }
    }
  }

  return item;
}

export function info(options: LogOptions) {
  return log(options, "info");
}

export function debug(options: LogOptions) {
  return log(options, "debug");
}

export function warn(options: LogOptions) {
  return log(options, "warn");
}

export function error(options: LogOptions) {
  return log(options, "error");
}
