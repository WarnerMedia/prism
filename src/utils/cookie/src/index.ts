import findTopDomain from "@segment/top-domain";
import { cookieHelper, debug } from "./helper";

export interface CookieModule {
  options(options?: CookieOptions): void | CookieOptions;
  set(name: string, value: string | object | boolean): boolean;
  get(name: string): string | object | boolean | null;
  remove(name: string): boolean;
}

export interface CookieOptions {
  maxage?: number;
  expires?: Date;
  domain?: string | null;
  path?: string;
  secure?: boolean;
  samesite?: string;
  useRootDomain?: boolean;
  encode?: boolean;
}

export class Cookie implements CookieModule {
  private _options: CookieOptions = {};

  constructor(options?: CookieOptions) {
    this.options = this.options.bind(this);
    this.set = this.set.bind(this);
    this.get = this.get.bind(this);
    this.getAll = this.getAll.bind(this);
    this.remove = this.remove.bind(this);
    this.options(options);
  }

  options(options?: CookieOptions): undefined | CookieOptions {
    if (arguments.length === 0) return this._options;

    options = options || {};

    let domain: CookieOptions["domain"] =
      options.domain || `.${findTopDomain(window.location.href)}`;

    if (domain === ".") {
      domain = null;
    }

    this._options = {
      maxage: 31536000000,
      path: "/",
      samesite: "Lax",
      encode: true,
      ...options,
      domain,
    };

    // http://curl.haxx.se/rfc/cookie_spec.html
    // https://publicsuffix.org/list/effective_tld_names.dat
    //
    // try setting a dummy cookie with the options
    // if the cookie isn't set, it probably means
    // that the domain is on the public suffix list
    // like myapp.deployment.com or localhost / ip.
    this.set("psm:test", true, this._options);
    if (!this.get("psm:test")) {
      debug(
        "[Cookie]: psm:test cookie could not be set at domain: ",
        this._options.domain
      );
      const reg = new RegExp(document.location.hostname, "g");
      if (document.referrer !== "" && document.referrer.search(reg) === -1) {
        this._options.samesite = "None";
      } else {
        this._options.domain = null;
      }
    }
    this.remove("psm:test");

    debug("[COOKIE]: this._options", this._options);
  }

  set(
    name: string,
    value: string | object | boolean,
    options: Partial<CookieOptions> = {}
  ): boolean {
    const opts = { ...this._options, ...options };

    debug("[COOKIE]: this.set() options", opts);

    try {
      if (typeof value !== "string") {
        value = JSON.stringify(value);
      }
      // @ts-ignore
      cookieHelper(name, value === "null" ? null : value, opts);
      return true;
    } catch (err) {
      debug("[COOKIE]: this.set() encountered an error", err);
      return false;
    }
  }

  get(name: string): string | object | boolean | null {
    const value = cookieHelper(name) as string | undefined;

    try {
      return JSON.parse(value);
    } catch (err) {
      // debug('[COOKIE]: this.get() encountered an error', err);
      return value || null;
    }
  }

  getAll() {
    return cookieHelper();
  }

  remove(name: string): boolean {
    try {
      cookieHelper(name, null, { ...this._options });
      return true;
    } catch (err) {
      debug("[COOKIE]: this.remove() encountered an error", err);
      return false;
    }
  }
}

export const cookie = new Cookie();
export { findTopDomain };
