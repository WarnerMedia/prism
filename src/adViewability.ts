// Copyright (c) Warner Media, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
import { isElementVisible } from "./utils/browser";

declare global {
  interface Window {
    googletag: googletag.Googletag;
  }
}

export interface SlotMetrics {
  slotId: string;
  slotSize: string;
  adUnitPath?: string;
  totalViewTime: number;
  lastViewStarted: number;
  visibleOnStart: boolean;
  adWasViewed: boolean;
  rfv?: string;
  creativeId?: string;
  lineItemId?: string;
  totalHoverTime: number;
  lastHoverStarted: number;
}

const win = window;
const doc = document;
let adObserver: IntersectionObserver;

export const registeredSlots: SlotMetrics[] = [];

win.googletag =
  win.googletag || ({ cmd: [] } as unknown as googletag.Googletag);

const log = (...args: string[]) => {
  if (win.location.search.search(/[?&]psm_debug=[1t]/) !== -1) {
    // console.log('[PSM]:', ...args);
  }
};

const timeInSeconds = () => Math.round(Date.now() / 1000);

const addDistinctRegisteredSlot = (slot: SlotMetrics) => {
  if (!registeredSlots.some((regSlot) => slot.slotId === regSlot.slotId)) {
    registeredSlots.push(slot);
  }
};

export function initAdViewListeners(): void {
  adObserver = new IntersectionObserver(
    (entries) => {
      entries.forEach((entry) => {
        const slotId = entry.target.id;
        const inViewPercentage = Math.round(entry.intersectionRatio) * 100;
        const rect = entry.boundingClientRect;

        if (entry.isIntersecting) {
          if (
            entry.intersectionRatio >= 0.5 &&
            isElementVisible(entry.target)
          ) {
            log(
              `slotVisibilityChanged - ${slotId} is in view. (${inViewPercentage}% in view)`
            );
            registeredSlots.forEach((slot) => {
              if (slot.slotId === slotId) {
                slot.visibleOnStart = true;
                slot.adWasViewed = true;
                slot.slotSize = `${rect.width}x${rect.height}`;
                updateSlotTimer(slot);
              }
            });
          }
        } else {
          // if an ad is not considered in view, flag it as not visible and reset lastViewStarted
          log(
            `slotVisibilityChanged - ${slotId} is no longer in view. (${inViewPercentage}% in view)`
          );
          registeredSlots.forEach((slot) => {
            if (slot.slotId === slotId) {
              slot.visibleOnStart = false;
              slot.lastViewStarted = 0;
              slot.lastHoverStarted = 0;
            }
          });
        }
      });
    },
    {
      root: null,
      rootMargin: "0px",
      threshold: [0.0, 0.5, 1.0],
    }
  );

  win.googletag.cmd.push(() => {
    win.googletag
      .pubads()
      .addEventListener(
        "slotRenderEnded",
        (e: googletag.events.SlotRenderEndedEvent) => {
          const slotId = e.slot.getSlotElementId();
          const adUnitPath = e.slot.getAdUnitPath();
          const elem = doc.getElementById(slotId);
          let slotSize = "0x0";

          if (e.size) {
            if (typeof e.size === "string") {
              slotSize = e.size;
            } else {
              slotSize = e.size.join("x");
            }
          }

          const slot: SlotMetrics = {
            slotId,
            slotSize,
            adUnitPath,
            totalViewTime: 0,
            lastViewStarted: 0,
            visibleOnStart: false,
            adWasViewed: false,
            rfv: e.slot.getTargeting("rfv")[0],
            creativeId: e.slot.getTargeting("creativeId")[0],
            lineItemId: e.slot.getTargeting("lineItemId")[0],
            totalHoverTime: 0,
            lastHoverStarted: 0,
          };

          addDistinctRegisteredSlot(slot);

          if (elem) {
            adObserver.observe(elem);
            adListeners(elem, slot);
          }
        }
      );

    win.googletag
      .pubads()
      .addEventListener(
        "slotVisibilityChanged",
        (e: googletag.events.SlotVisibilityChangedEvent) => {
          const slotId = e.slot.getSlotElementId();
          registeredSlots.forEach((slot) => {
            if (slot.slotId === slotId) {
              slot.rfv = e.slot.getTargeting("rfv")[0];
              (slot.creativeId = e.slot.getTargeting("creativeId")[0]),
                (slot.lineItemId = e.slot.getTargeting("lineItemId")[0]);
            }
          });
        }
      );

    win.googletag
      .pubads()
      .getSlots()
      .forEach((slot: googletag.Slot) => {
        const slotId = slot.getSlotElementId();
        const elem = doc.getElementById(slotId);
        let rect = {
          width: 0,
          height: 0,
        };

        if (elem) {
          rect = elem.getBoundingClientRect();
        }

        const adSlot: SlotMetrics = {
          slotId,
          slotSize: rect
            ? `${Math.round(rect.width)}x${Math.round(rect.height)}`
            : "0x0",
          adUnitPath: slot.getAdUnitPath(),
          totalViewTime: 0,
          lastViewStarted: 0,
          visibleOnStart: false,
          adWasViewed: false,
          rfv: slot.getTargeting("rfv")[0],
          creativeId: slot.getTargeting("creativeId")[0],
          lineItemId: slot.getTargeting("lineItemId")[0],
          totalHoverTime: 0,
          lastHoverStarted: 0,
        };

        addDistinctRegisteredSlot(adSlot);

        if (elem) {
          adObserver.observe(elem);
          adListeners(elem, adSlot);
        }
      });
  });

  // listen for the inbrain.loaded event and retrieve array of DOM IDs
  // each Inbrain promo item should be registered as a distinct ad slot
  doc.addEventListener("inbrain.loaded", ((e: CustomEvent) => {
    e.detail.forEach((target) => {
      const elem = doc.getElementById(target);

      if (elem) {
        const rect = elem.getBoundingClientRect();

        addDistinctRegisteredSlot({
          slotId: target,
          slotSize: rect
            ? `${Math.round(rect.width)}x${Math.round(rect.height)}`
            : "0x0",
          totalViewTime: 0,
          lastViewStarted: 0,
          visibleOnStart: false,
          adWasViewed: false,
          totalHoverTime: 0,
          lastHoverStarted: 0,
        });

        adObserver.observe(elem);
      }
    });
  }) as EventListener);
}

export function onHiddenAds() {
  // remove ad slots from the set of visible slots and mark them as inactive
  // so we can restore them when a user tabs back into the document
  registeredSlots.forEach((slot) => {
    if (slot.visibleOnStart) {
      slot.lastViewStarted = 0;
    }
    const elem = doc.getElementById(slot.slotId);

    if (elem) {
      adObserver.unobserve(elem);
    }
  });
}

export function onVisibleAds() {
  // reset the last view started to the current time when the document first becomes visible
  registeredSlots.forEach((slot) => {
    if (slot.visibleOnStart) {
      slot.lastViewStarted = timeInSeconds();
    }
    const elem = doc.getElementById(slot.slotId);

    if (elem) {
      adObserver.observe(elem);
    }
  });
}

export function setAdViewMetrics() {
  registeredSlots.forEach((slot) => {
    if (slot.visibleOnStart) {
      updateSlotTimer(slot);
    }
  });
}

export function resetAdViewMetrics() {
  registeredSlots.forEach((slot) => {
    slot.totalViewTime = 0;
    slot.totalHoverTime = 0;
    slot.adWasViewed = false;
  });
}

function updateSlotTimer(slot: SlotMetrics) {
  const lastStarted = slot.lastViewStarted;
  const currentTime = timeInSeconds();

  // if the timer is currently running
  if (lastStarted) {
    // compute the difference between the current time and the start time
    // to determine the number of seconds the ad slot has been visible since the last
    // time it became visible
    const diff = currentTime - lastStarted;

    slot.totalViewTime += diff;
  }

  // update the last viewed time to the current time
  slot.lastViewStarted = currentTime;
}

function updateAdHoverTimer(event: string, slot: SlotMetrics) {
  const currentTime = timeInSeconds();
  let diff = 0;

  if (slot.lastHoverStarted !== 0) {
    diff = currentTime - slot.lastHoverStarted;
  }

  if (event === "mouseleave") {
    slot.totalHoverTime += diff;
  } else {
    slot.lastHoverStarted = currentTime;
  }
}

function adListeners(target: HTMLElement, slot: SlotMetrics): void {
  target.addEventListener("mouseenter", () => {
    updateAdHoverTimer("mouseenter", slot);
  });

  target.addEventListener("mouseleave", () => {
    updateAdHoverTimer("mouseleave", slot);
  });
}
