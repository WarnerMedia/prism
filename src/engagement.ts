// Copyright (c) Warner Media, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
import { EngagementMetrics } from "./payloadCore";
import { detectDocumentSize, throttle } from "./utils/browser";
import { HEARTBEAT_INTERVAL } from "./utils/constants";

const win = window;
const doc = document;

const documentEvents = ["mousedown", "mousemove", "keydown"];
const windowEvents = ["scroll", "focus"];

export let lastEngagementEvent: number;
export let hbStartTimestamp: string = new Date().toISOString();

export let engagement: EngagementMetrics = {
  totalTime: 0,
  engagedTime: 0,
  interval: HEARTBEAT_INTERVAL / 1000,
  maxScrollDepth: win.scrollY < 0 ? 0 : win.scrollY,
  currentScrollPosition: win.scrollY < 0 ? 0 : win.scrollY,
  didScrollUpDuringInterval: false,
};

function handleEngagementEvent() {
  const current = win.scrollY < 0 ? 0 : win.scrollY;
  const totalHeight = detectDocumentSize()[1];

  if (current > engagement.currentScrollPosition) {
    engagement.maxScrollDepth = current >= totalHeight ? totalHeight : current;
  }
  if (current < engagement.currentScrollPosition) {
    engagement.didScrollUpDuringInterval = true;
  }
  engagement.currentScrollPosition = current;

  const now = new Date();

  // if we have a timestamp that denotes the last engagement event
  if (lastEngagementEvent) {
    // calculate the difference between the current time and the last engagement event in seconds
    const diff = Math.round((now.getTime() - lastEngagementEvent) / 1000);

    // if the difference is within a 3 second window, consider the difference "engaged" time
    // and add it to the engaged time total
    if (diff <= 3) {
      engagement.engagedTime += diff;
    }
  }

  // reset the last engagement event time to the current time
  lastEngagementEvent = now.getTime();
}

export function initEngagementEvents() {
  documentEvents.forEach((event) => {
    doc.addEventListener(event, throttle(handleEngagementEvent, 1000));
  });

  windowEvents.forEach((event) => {
    win.addEventListener(event, throttle(handleEngagementEvent, 1000));
  });
}

export function setEngagementMetrics() {
  const now = new Date();
  const start = new Date(hbStartTimestamp);

  engagement.totalTime = Math.round((now.getTime() - start.getTime()) / 1000);
}

export function resetHbInterval() {
  const scroll = win.scrollY < 0 ? 0 : win.scrollY;
  engagement = {
    totalTime: 0,
    engagedTime: 0,
    interval: HEARTBEAT_INTERVAL / 1000,
    maxScrollDepth: scroll,
    currentScrollPosition: scroll,
    didScrollUpDuringInterval: false,
  };
  hbStartTimestamp = new Date().toISOString();
  lastEngagementEvent = Date.now();
}
