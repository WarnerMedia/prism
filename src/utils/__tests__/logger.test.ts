import { PsmConfig, Psm } from "../../Psm";
import * as logger from "../logger";
import {
  clearMockDate,
  mockDate,
  mockStorage,
  isoDate,
  mockUserAgent,
  clearMockUserAgent,
} from "../testUtils";
import { cookie } from "../cookie/src";

jest.mock("../../../package.json", () => ({
  version: "0.0.0",
}));

jest.mock("../sendRequest", () => ({
  sendRequest: jest.fn().mockResolvedValue({
    asn: { id: "1111", name: "example llc" },
    continent: "NA",
    continentName: "North America",
    country: "US",
    country_alpha2: "US",
    country_alpha3: "USA",
    ip_address: "192.0.2.0",
    lat: "50.00",
    lon: "-50.00",
    proxy: null,
    states: [
      { cities: ["ATLANTA"], counties: [], state: "GA", zipcodes: ["33333"] },
    ],
  }),
}));

jest.mock("../../privacy", () => ({
  ...(jest.requireActual("../../privacy") as object),
  checkOutsideUS: jest
    .fn()
    .mockResolvedValue([
      { "special-purpose-1": true },
      { "special-purpose-2": true },
      { "feature-1": true },
      { "feature-2": true },
      { "feature-3": true },
    ]),
}));

const err = {
  message: "test error",
  stack: "heres our stack",
  context: {
    requestUrl: "http://psm.test",
    requestHTTPStatus: 208302834,
    requestPayload: "payload",
    requestHeaders: "headers",
    requestMethod: "POST",
    responsePayload: "{}",
  },
};

describe("logger", () => {
  const validLogOptions: Partial<logger.LogOptions> = {
    methodName: "testMethod",
    eventType: "whatever",
  };
  const psm: Partial<Psm> = {
    ready: true,
    config: {
      brand: "Prism",
      subBrand: "",
      cookieDomain: ".psm.test",
      cookieExpires: 31536000000,
      cookieSameSite: "Lax",
      cookieSecure: false,
      psmEnvironment: "TEST",
      countryCode: "US",
      platform: "web",
    },
    getUKID() {
      return "00000000-0000-0000-0000-000000000000";
    },
    getCSID() {
      return "00000000-0000-0000-0000-000000000000";
    },
    getFlags() {
      return [
        {
          enabled: true,
          flagId: "consent-update",
          flagName: "consent-update",
        },
        {
          enabled: false,
          flagId: "inbrain",
          flagName: "inbrain",
        },
        {
          enabled: true,
          flagId: "heartbeat-event",
          flagName: "heartbeat-event",
        },
        {
          enabled: true,
          flagId: "pubsub-event",
          flagName: "pubsub-event",
        },
        {
          enabled: true,
          flagId: "outside-us-location-check",
          flagName: "outside-us-location-check",
        },
        {
          enabled: true,
          flagId: "identity-onstart",
          flagName: "identity-onstart",
        },
        {
          enabled: true,
          flagId: "identity-oncomplete",
          flagName: "identity-oncomplete",
        },
        {
          enabled: true,
          flagId: "idresolve",
          flagName: "idresolve",
        },
      ];
    },
    session: {
      pageloadid: 1,
      sessionid: "00000000-0000-0000-0000-000000000000",
      sessionDuration: 0,
      psmLastActiveTimestamp: isoDate,
      psmSessionStart: isoDate,
      isSessionStart: false,
    },
  };

  beforeAll(() => {
    cookie.set("UKID", "00000000-0000-0000-0000-000000000000");
  });

  beforeEach(() => {
    mockStorage();
    mockDate();
    mockUserAgent();
  });

  afterEach(() => {
    clearMockDate();
    clearMockUserAgent();
  });

  it("should log", () => {
    logger.createLogger(psm as Psm);

    const msg = logger.log(
      {
        err,
        ...validLogOptions,
      } as logger.LogOptions,
      "test"
    );

    expect(msg).toMatchSnapshot();
  });

  it("should debug", () => {
    logger.createLogger(psm as Psm);

    const msg = logger.debug({
      err,
      ...validLogOptions,
    } as logger.LogOptions);

    expect(msg).toMatchSnapshot();
  });

  it("should warn", () => {
    logger.createLogger(psm as Psm);

    const msg = logger.warn({
      err,
      ...validLogOptions,
    } as logger.LogOptions);

    expect(msg).toMatchSnapshot();
  });

  it("should error", () => {
    logger.createLogger(psm as Psm);

    const msg = logger.error({
      err,
      ...validLogOptions,
    } as logger.LogOptions);

    expect(msg).toMatchSnapshot();
  });
});
