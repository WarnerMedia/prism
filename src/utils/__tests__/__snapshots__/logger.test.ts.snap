// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger should debug 1`] = `
Object {
  "brand": "Prism",
  "device": {"type":""},
  "eventName": "debug",
  "eventProperties": Object {
    "featureFlagValues": Array [
      Object {
        "enabled": true,
        "flagId": "consent-update",
        "flagName": "consent-update",
      },
      Object {
        "enabled": false,
        "flagId": "inbrain",
        "flagName": "inbrain",
      },
      Object {
        "enabled": true,
        "flagId": "heartbeat-event",
        "flagName": "heartbeat-event",
      },
      Object {
        "enabled": true,
        "flagId": "pubsub-event",
        "flagName": "pubsub-event",
      },
      Object {
        "enabled": true,
        "flagId": "outside-us-location-check",
        "flagName": "outside-us-location-check",
      },
      Object {
        "enabled": true,
        "flagId": "identity-onstart",
        "flagName": "identity-onstart",
      },
      Object {
        "enabled": true,
        "flagId": "identity-oncomplete",
        "flagName": "identity-oncomplete",
      },
      Object {
        "enabled": true,
        "flagId": "idresolve",
        "flagName": "idresolve",
      },
    ],
    "logMessage": "[object Object]",
    "method": "testMethod",
    "stackTrace": "heres our stack",
  },
  "eventTimestamp": "2024-01-01T09:09:09.000Z",
  "eventType": "log",
  "library": Object {
    "initConfig": Object {
      "brand": "Prism",
      "cookieDomain": ".psm.test",
      "cookieExpires": 31536000000,
      "cookieSameSite": "Lax",
      "cookieSecure": false,
      "countryCode": "US",
      "platform": "web",
      "psmEnvironment": "TEST",
      "subBrand": "",
    },
    "name": "PrismJS",
    "version": "0.0.0",
  },
  "platform": "web",
  "session": Object {
    "isSessionStart": false,
    "pageloadid": 1,
    "psmLastActiveTimestamp": "2024-01-01T09:09:09.000Z",
    "psmSessionStart": "2024-01-01T09:09:09.000Z",
    "sessionDuration": 0,
    "sessionid": "00000000-0000-0000-0000-000000000000",
  },
  "ukid": undefined,
}
`;

exports[`logger should error 1`] = `
Object {
  "brand": "Prism",
  "device": {"type":""},
  "eventName": "error",
  "eventProperties": Object {
    "featureFlagValues": Array [
      Object {
        "enabled": true,
        "flagId": "consent-update",
        "flagName": "consent-update",
      },
      Object {
        "enabled": false,
        "flagId": "inbrain",
        "flagName": "inbrain",
      },
      Object {
        "enabled": true,
        "flagId": "heartbeat-event",
        "flagName": "heartbeat-event",
      },
      Object {
        "enabled": true,
        "flagId": "pubsub-event",
        "flagName": "pubsub-event",
      },
      Object {
        "enabled": true,
        "flagId": "outside-us-location-check",
        "flagName": "outside-us-location-check",
      },
      Object {
        "enabled": true,
        "flagId": "identity-onstart",
        "flagName": "identity-onstart",
      },
      Object {
        "enabled": true,
        "flagId": "identity-oncomplete",
        "flagName": "identity-oncomplete",
      },
      Object {
        "enabled": true,
        "flagId": "idresolve",
        "flagName": "idresolve",
      },
    ],
    "logMessage": "[object Object]",
    "method": "testMethod",
    "stackTrace": "heres our stack",
  },
  "eventTimestamp": "2024-01-01T09:09:09.000Z",
  "eventType": "log",
  "library": Object {
    "initConfig": Object {
      "brand": "Prism",
      "cookieDomain": ".psm.test",
      "cookieExpires": 31536000000,
      "cookieSameSite": "Lax",
      "cookieSecure": false,
      "countryCode": "US",
      "platform": "web",
      "psmEnvironment": "TEST",
      "subBrand": "",
    },
    "name": "PrismJS",
    "version": "0.0.0",
  },
  "platform": "web",
  "session": Object {
    "isSessionStart": false,
    "pageloadid": 1,
    "psmLastActiveTimestamp": "2024-01-01T09:09:09.000Z",
    "psmSessionStart": "2024-01-01T09:09:09.000Z",
    "sessionDuration": 0,
    "sessionid": "00000000-0000-0000-0000-000000000000",
  },
  "ukid": undefined,
}
`;

exports[`logger should log 1`] = `
Object {
  "brand": "Prism",
  "device": {"type":""},
  "eventName": "test",
  "eventProperties": Object {
    "featureFlagValues": Array [
      Object {
        "enabled": true,
        "flagId": "consent-update",
        "flagName": "consent-update",
      },
      Object {
        "enabled": false,
        "flagId": "inbrain",
        "flagName": "inbrain",
      },
      Object {
        "enabled": true,
        "flagId": "heartbeat-event",
        "flagName": "heartbeat-event",
      },
      Object {
        "enabled": true,
        "flagId": "pubsub-event",
        "flagName": "pubsub-event",
      },
      Object {
        "enabled": true,
        "flagId": "outside-us-location-check",
        "flagName": "outside-us-location-check",
      },
      Object {
        "enabled": true,
        "flagId": "identity-onstart",
        "flagName": "identity-onstart",
      },
      Object {
        "enabled": true,
        "flagId": "identity-oncomplete",
        "flagName": "identity-oncomplete",
      },
      Object {
        "enabled": true,
        "flagId": "idresolve",
        "flagName": "idresolve",
      },
    ],
    "logMessage": "[object Object]",
    "method": "testMethod",
    "stackTrace": "heres our stack",
  },
  "eventTimestamp": "2024-01-01T09:09:09.000Z",
  "eventType": "log",
  "library": Object {
    "initConfig": Object {
      "brand": "Prism",
      "cookieDomain": ".psm.test",
      "cookieExpires": 31536000000,
      "cookieSameSite": "Lax",
      "cookieSecure": false,
      "countryCode": "US",
      "platform": "web",
      "psmEnvironment": "TEST",
      "subBrand": "",
    },
    "name": "PrismJS",
    "version": "0.0.0",
  },
  "platform": "web",
  "session": Object {
    "isSessionStart": false,
    "pageloadid": 1,
    "psmLastActiveTimestamp": "2024-01-01T09:09:09.000Z",
    "psmSessionStart": "2024-01-01T09:09:09.000Z",
    "sessionDuration": 0,
    "sessionid": "00000000-0000-0000-0000-000000000000",
  },
  "ukid": undefined,
}
`;

exports[`logger should warn 1`] = `
Object {
  "brand": "Prism",
  "device": {"type":""},
  "eventName": "warn",
  "eventProperties": Object {
    "featureFlagValues": Array [
      Object {
        "enabled": true,
        "flagId": "consent-update",
        "flagName": "consent-update",
      },
      Object {
        "enabled": false,
        "flagId": "inbrain",
        "flagName": "inbrain",
      },
      Object {
        "enabled": true,
        "flagId": "heartbeat-event",
        "flagName": "heartbeat-event",
      },
      Object {
        "enabled": true,
        "flagId": "pubsub-event",
        "flagName": "pubsub-event",
      },
      Object {
        "enabled": true,
        "flagId": "outside-us-location-check",
        "flagName": "outside-us-location-check",
      },
      Object {
        "enabled": true,
        "flagId": "identity-onstart",
        "flagName": "identity-onstart",
      },
      Object {
        "enabled": true,
        "flagId": "identity-oncomplete",
        "flagName": "identity-oncomplete",
      },
      Object {
        "enabled": true,
        "flagId": "idresolve",
        "flagName": "idresolve",
      },
    ],
    "logMessage": "[object Object]",
    "method": "testMethod",
    "stackTrace": "heres our stack",
  },
  "eventTimestamp": "2024-01-01T09:09:09.000Z",
  "eventType": "log",
  "library": Object {
    "initConfig": Object {
      "brand": "Prism",
      "cookieDomain": ".psm.test",
      "cookieExpires": 31536000000,
      "cookieSameSite": "Lax",
      "cookieSecure": false,
      "countryCode": "US",
      "platform": "web",
      "psmEnvironment": "TEST",
      "subBrand": "",
    },
    "name": "PrismJS",
    "version": "0.0.0",
  },
  "platform": "web",
  "session": Object {
    "isSessionStart": false,
    "pageloadid": 1,
    "psmLastActiveTimestamp": "2024-01-01T09:09:09.000Z",
    "psmSessionStart": "2024-01-01T09:09:09.000Z",
    "sessionDuration": 0,
    "sessionid": "00000000-0000-0000-0000-000000000000",
  },
  "ukid": undefined,
}
`;
